### PR TITLE
Fix worker reordering stack overflow

### PR DESCRIPTION
**Pull request by matthiakl**
Hotfix for https://github.com/widelands/widelands/issues/4995#issuecomment-972008401

The problem is, that the reordering does not terminate for the european basic farm, where both slots need the same worker.
This quick fix solves the issue, but I will look, whether there is a better solution.

### CODE REVIEW

**Comment by klaus-halfmann** [comment]
Code looks straight forward, AFAIK I need the Europeans tribe. What/How I can I test there?

**Comment by klaus-halfmann** [comment]
Is it normal that Europeans result in such ugly color overlays?

grafik

**Comment by Noordfrees** [comment]
> Is it normal that Europeans result in such ugly color overlays?
- That's #4881

File: src/logic/map_objects/tribes/productionsite.cc [hunk 1]
```diff
@@ -627,7 +627,13 @@ std::pair<int32_t, int32_t> ProductionSite::find_worker(OPtr<Worker> w) {
  */
 void ProductionSite::try_replace_worker(const Game* game,
                                         DescriptionIndex worker_index,
-                                        WorkingPosition* wp) {
+                                        WorkingPosition* wp,
+                                        size_t iteration) {
+	if (iteration >= working_positions_.size()) {
+		// cancel recursion, something went wrong
+		log_warn("Aborting try_replace_worker() after %" PRIuS " calls.", iteration);
```
Comment (hessenfarmer) [inline src/logic/map_objects/tribes/productionsite.cc]: this should be made verbose to declutter the log for mat5ches with europeans or other future tribes having similar designed productionsites

**Comment by matthiakl** [comment]
Now it is a proper solution. There was an issue with the `needs_higher_qualification` condition, which also matched "same qualification"

**Review by hessenfarmer — APPROVED** [review]
code looks straight forward.
@MarkMcWire does the thumb mean you tested this? 
if yes this can go in

**Comment by MarkMcWire** [comment]
> code looks straight forward. @MarkMcWire does the thumb mean you tested this? if yes this can go in

I will test it today. Yesterday, was just a "well done".

**Comment by MarkMcWire** [comment]
> code looks straight forward. @MarkMcWire does the thumb mean you tested this? if yes this can go in

-> https://drive.google.com/file/d/1DDHR-8C1gbCYZ1wkYX_h8Ps45ZcoCs25/view?usp=sharing
_The game don't crash anymore_, but the performance is slower than without the fix. The game "jerks" a little.

It crashes for another reason:

```
widelands: /home/solus/Git/widelands_fix/widelands/src/wui/inputqueuedisplay.cc:351: int32_t InputQueueDisplay::fill_index_at(int32_t, int32_t) const: Zusicherung »nr_icons_ > 0« nicht erfüllt.
Abgebrochen (Speicherabzug geschrieben)
solus@solus-pc ~/Git/widelands_fix/widelands
```

**Comment by hessenfarmer** [comment]
> 
> 
> > code looks straight forward. @MarkMcWire does the thumb mean you tested this? if yes this can go in
> 
> -> https://drive.google.com/file/d/1DDHR-8C1gbCYZ1wkYX_h8Ps45ZcoCs25/view?usp=sharing _The game don't crash anymore_, but the performance is slower than without the fix. The game "jerks" a little.
   
Well it is hard to believe that such a small fix has any influence on the performance except in a very big economy. The short freezes of the display are present in master as well and are most probably caused by the AI think cycles or routing se #4963 as the efeects of the mentioned bug are sometimes more visible then other times I believe this might have led to the impression it may be related to this PR.

> 
> It crashes for another reason:
> 
> ```
> widelands: /home/solus/Git/widelands_fix/widelands/src/wui/inputqueuedisplay.cc:351: int32_t InputQueueDisplay::fill_index_at(int32_t, int32_t) const: Zusicherung »nr_icons_ > 0« nicht erfüllt.
> Abgebrochen (Speicherabzug geschrieben)
> solus@solus-pc ~/Git/widelands_fix/widelands
> ```

this is almost 100% another flavour of the famous #5052 bug

So as the original crash is gone I think we should have this fix in

**Pull request merged**